### ISSUE TITLE
improve compile error when using comptime concatenation and operand has runtime known length

### ISSUE THREAD

**Issue by nektro** [ziglang/zig#11773]
### Zig Version

0.10.0-dev.2444+83beed09e

### Steps to Reproduce

```zig
const std = @import("std");

test {
    const a = foo() ++ "bb";
    try std.testing.expect(a.len == 5);
}

fn foo() []const u8 {
    return "aaa";
}
```

### Expected Behavior

```
All 1 tests passed.
```

### Actual Behavior

```
test.zig:4:18: error: unable to resolve comptime value
    const a = foo() ++ "bb";
                 ^
```

only affects stage2

**Comment by nektro** [comment]
moving `a` to decl scope makes this work

**Comment by Vexu** [comment]
Not a bug, stage2 implements #7147 so the operands aren't implicitly conptime but slice concatenation still requires the length to be known.

**Comment by andrewrk** [comment]
Yep this is working as designed.

**Comment by andrewrk** [comment]
Re-opening for a better compile error. It should say something like, "array concatenation operand requires comptime-known length".

**Comment by Vexu** [comment]
Fixed in #12117
```
a.zig:6:18: error: unable to resolve comptime value
    const a = foo() ++ "bb";
              ~~~^~
a.zig:6:18: note: slice value being concatenated must be comptime-known
```
